### PR TITLE
Calculate Heartbeat's timestamp on receiving

### LINKED ISSUE #1610: Incorrectly marking agents as crashed
1. Agent sends `FINISHED` heartbeat
2. Orchestrator has an exception in `agentService.checkSavedData`
3. Latest heartbeat timestamp isn't updated in `HeartbeatInspector`
4. After some time, `HeartbeatInspector` initiates destruction of this agent

Agent should keep trying to send this heartbeat until orchestrator is able to accept it.

Version: e571ecc96b42d052930741a7daf40154db89a034

<details>
  <summary>Logs:</summary>

Orchestrator:
```
08:20:05.344 [reactor-http-epoll-7] INFO c.s.s.o.c.HeartbeatController - Got heartbeat state: FINISHED from save-execution-8-mdwb8 under execution id=8
08:20:05.450 [reactor-http-epoll-6] ERROR o.s.b.a.w.r.e.AbstractErrorWebExceptionHandler - [a9492ece-506]  500 Server Error for HTTP POST "/heartbeat" org.springframework.web.reactive.function.client.WebClientResponseException$NotFound: 404 Not Found from GET http://backend/internal/testExecutions/agent/save-execution-8-mdwb8/READY_FOR_TESTING
	at <...>
08:20:05.460 [reactor-http-epoll-6] DEBUG o.s.h.codec.json.Jackson2JsonEncoder - [a9492ece-506] Encoding [{timestamp=Wed Dec 14 08:20:05 UTC 2022, path=/heartbeat, status=500, error=Internal Server Error, r (truncated)...] 
08:20:05.466 [reactor-http-epoll-7] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [a9492ece-506] Completed 500 INTERNAL_SERVER_ERROR

08:20:08.377 [reactor-http-epoll-7] INFO c.s.s.o.c.HeartbeatController - Got heartbeat state: FINISHED from save-execution-8-mdwb8 under execution id=8
08:20:08.432 [reactor-http-epoll-6] ERROR o.s.b.a.w.r.e.AbstractErrorWebExceptionHandler - [a9492ece-507]  500 Server Error for HTTP POST "/heartbeat" org.springframework.web.reactive.function.client.WebClientResponseException$NotFound: 404 Not Found from GET http://backend/internal/testExecutions/agent/save-execution-8-mdwb8/READY_FOR_TESTING
	at <...>
08:20:08.433 [reactor-http-epoll-6] DEBUG o.s.h.codec.json.Jackson2JsonEncoder - [a9492ece-507] Encoding [{timestamp=Wed Dec 14 08:20:08 UTC 2022, path=/heartbeat, status=500, error=Internal Server Error, r (truncated)...] 
08:20:08.434 [reactor-http-epoll-7] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [a9492ece-507] Completed 500 INTERNAL_SERVER_ERROR 
08:20:10.001 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Latest heartbeat from save-execution-8-mdwb8 was sent: 4663 ms ago 
08:20:10.014 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Pod name=save-execution-8-mdwb8 is still present; reason=Running, isRunning=true, conditions=[PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=Initialized, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=Ready, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=ContainersReady, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=PodScheduled, additionalProperties={})] 

08:20:12.461 [reactor-http-epoll-7] INFO c.s.s.o.c.HeartbeatController - Got heartbeat state: FINISHED from save-execution-8-mdwb8 under execution id=8
08:20:12.517 [reactor-http-epoll-6] ERROR o.s.b.a.w.r.e.AbstractErrorWebExceptionHandler - [a9492ece-508]  500 Server Error for HTTP POST "/heartbeat" org.springframework.web.reactive.function.client.WebClientResponseException$NotFound: 404 Not Found from GET http://backend/internal/testExecutions/agent/save-execution-8-mdwb8/READY_FOR_TESTING
	at <...>
08:20:12.518 [reactor-http-epoll-6] DEBUG o.s.h.codec.json.Jackson2JsonEncoder - [a9492ece-508] Encoding [{timestamp=Wed Dec 14 08:20:12 UTC 2022, path=/heartbeat, status=500, error=Internal Server Error, r (truncated)...] 
08:20:12.519 [reactor-http-epoll-7] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [a9492ece-508] Completed 500 INTERNAL_SERVER_ERROR 

08:20:21.301 [reactor-http-epoll-7] INFO c.s.s.o.c.HeartbeatController - Got heartbeat state: FINISHED from save-execution-8-mdwb8 under execution id=8
08:20:21.366 [reactor-http-epoll-6] ERROR o.s.b.a.w.r.e.AbstractErrorWebExceptionHandler - [a9492ece-511]  500 Server Error for HTTP POST "/heartbeat" org.springframework.web.reactive.function.client.WebClientResponseException$NotFound: 404 Not Found from GET http://backend/internal/testExecutions/agent/save-execution-8-mdwb8/READY_FOR_TESTING
	at <...>
08:20:21.367 [reactor-http-epoll-6] DEBUG o.s.h.codec.json.Jackson2JsonEncoder - [a9492ece-511] Encoding [{timestamp=Wed Dec 14 08:20:21 UTC 2022, path=/heartbeat, status=500, error=Internal Server Error, r (truncated)...] 
08:20:21.368 [reactor-http-epoll-7] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [a9492ece-511] Completed 500 INTERNAL_SERVER_ERROR 

08:20:37.841 [reactor-http-epoll-7] INFO c.s.s.o.c.HeartbeatController - Got heartbeat state: FINISHED from save-execution-8-mdwb8 under execution id=8 
08:20:37.899 [reactor-http-epoll-6] ERROR o.s.b.a.w.r.e.AbstractErrorWebExceptionHandler - [a9492ece-516]  500 Server Error for HTTP POST "/heartbeat" org.springframework.web.reactive.function.client.WebClientResponseException$NotFound: 404 Not Found from GET http://backend/internal/testExecutions/agent/save-execution-8-mdwb8/READY_FOR_TESTING
	at <...>
08:20:37.900 [reactor-http-epoll-6] DEBUG o.s.h.codec.json.Jackson2JsonEncoder - [a9492ece-516] Encoding [{timestamp=Wed Dec 14 08:20:37 UTC 2022, path=/heartbeat, status=500, error=Internal Server Error, r (truncated)...] 
08:20:37.901 [reactor-http-epoll-7] DEBUG o.s.w.s.a.HttpWebHandlerAdapter - [a9492ece-516] Completed 500 INTERNAL_SERVER_ERROR 
08:20:40.000 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Latest heartbeat from save-execution-8-mdwb8 was sent: 34662 ms ago 
08:20:40.013 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Pod name=save-execution-8-mdwb8 is still present; reason=Running, isRunning=true, conditions=[PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=Initialized, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=Ready, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=ContainersReady, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=PodScheduled, additionalProperties={})] 

08:20:50.001 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Latest heartbeat from save-execution-8-mdwb8 was sent: 44662 ms ago 
08:20:50.017 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Pod name=save-execution-8-mdwb8 is still present; reason=Running, isRunning=true, conditions=[PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=Initialized, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=Ready, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=ContainersReady, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=PodScheduled, additionalProperties={})] 

08:21:00.000 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Latest heartbeat from save-execution-8-mdwb8 was sent: 54662 ms ago 
08:21:00.017 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Pod name=save-execution-8-mdwb8 is still present; reason=Running, isRunning=true, conditions=[PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=Initialized, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=Ready, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=ContainersReady, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=PodScheduled, additionalProperties={})] 

08:21:10.001 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Latest heartbeat from save-execution-8-mdwb8 was sent: 64663 ms ago 
08:21:10.001 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Adding save-execution-8-mdwb8 to list crashed agents 
08:21:10.025 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Pod name=save-execution-8-mdwb8 is still present; reason=Running, isRunning=true, conditions=[PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=Initialized, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=Ready, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=ContainersReady, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=PodScheduled, additionalProperties={})] 
08:21:10.031 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Pod name=save-execution-8-mdwb8 is still present; reason=Running, isRunning=true, conditions=[PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=Initialized, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=Ready, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:55Z, message=null, reason=null, status=True, type=ContainersReady, additionalProperties={}), PodCondition(lastProbeTime=null, lastTransitionTime=2022-12-14T08:14:52Z, message=null, reason=null, status=True, type=PodScheduled, additionalProperties={})] 
08:21:10.031 [scheduling-1] DEBUG c.s.s.o.service.HeartBeatInspector - Stopping crashed agents: [save-execution-8-mdwb8] 
08:21:10.110 [scheduling-1] DEBUG c.s.s.o.kubernetes.KubernetesManager - Deleted pod with name=save-execution-8-mdwb8 
08:21:10.152 [scheduling-1] WARN c.s.s.o.service.HeartBeatInspector - All agents are crashed, initialize shutdown sequence. Crashed agents: [save-execution-8-mdwb8]

08:21:15.456 [reactor-http-epoll-6] DEBUG c.s.s.o.service.AgentService - For executionId=8 agent statuses are [AgentStatusDto(state=CRASHED, containerId=save-execution-8-mdwb8, time=2022-12-14T08:21:10.112)] 
08:21:15.456 [reactor-http-epoll-6] DEBUG c.s.s.o.service.AgentService - For execution id=8 there are finished or stopped agents 
08:21:15.456 [reactor-http-epoll-6] INFO c.s.s.o.service.AgentService - For execution id=8 all agents have completed their lifecycle 
08:21:15.458 [reactor-http-epoll-6] DEBUG c.s.s.o.kubernetes.KubernetesManager - Removing a Job for execution id=8 
```

Agent:
```
[INFO]: [tid 11] [HTTP Client] REQUEST: http://orchestrator/heartbeat
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:20:05.337987600Z"}
{"timestamp":"2022-12-14T08:20:05.446+00:00","path":"/heartbeat","status":500,"error":"Internal Server Error","requestId":"a9492ece-506"}

[DEBUG]: [tid 11] Retrying request: attempt #2, reason: 500 Internal Server Error
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:20:05.337987600Z"}
{"timestamp":"2022-12-14T08:20:12.517+00:00","path":"/heartbeat","status":500,"error":"Internal Server Error","requestId":"a9492ece-508"}

[DEBUG]: [tid 11] Retrying request: attempt #3, reason: 500 Internal Server Error
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:20:05.337987600Z"}
{"timestamp":"2022-12-14T08:20:21.365+00:00","path":"/heartbeat","status":500,"error":"Internal Server Error","requestId":"a9492ece-511"}

[DEBUG]: [tid 11] Retrying request: attempt #4, reason: 500 Internal Server Error
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:20:05.337987600Z"}
{"timestamp":"2022-12-14T08:20:37.898+00:00","path":"/heartbeat","status":500,"error":"Internal Server Error","requestId":"a9492ece-516"}

[DEBUG]: [tid 11] Retrying request: attempt #5, reason: 500 Internal Server Error
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:20:05.337987600Z"}
[INFO]: [tid 11] [HTTP Client] REQUEST http://orchestrator/heartbeat failed with exception: kotlin.coroutines.cancellation.CancellationException: Request timeout has expired [url=http://orchestrator/heartbeat, request_timeout=60000 ms]
[INFO]: [tid 11] Waiting for 15000 ms

[INFO]: [tid 11] [HTTP Client] REQUEST: http://orchestrator/heartbeat
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:21:25.335797800Z"}
{"timestamp":"2022-12-14T08:21:25.392+00:00","path":"/heartbeat","status":500,"error":"Internal Server Error","requestId":"a9492ece-527"}

[DEBUG]: [tid 11] Retrying request: attempt #2, reason: 500 Internal Server Error
{"containerId":"save-execution-8-mdwb8","state":"FINISHED","executionProgress":{"percentCompletion":0,"executionId":8},"timestamp":"2022-12-14T08:21:25.335797800Z"}
{"timestamp":"2022-12-14T08:21:31.773+00:00","path":"/heartbeat","status":500,"error":"Internal Server Error","requestId":"a9492ece-529"}
```
</details>

So there was a heartbeat at `08:20:37.898`, but then we see `Latest heartbeat from save-execution-8-mdwb8 was sent: **34662** ms ago` at `08:20:40.000`, i.e. just two seconds later.

Before that, though, timestamp is correctly updated at `08:20:05.344` even though there is `500` once again.

---

Okay, that's the reason:
```kotlin
    fun updateAgentHeartbeatTimeStamps(heartbeat: Heartbeat) {
        agentsLatestHeartBeatsMap[heartbeat.containerId] = heartbeat.state.name to heartbeat.timestamp
    }
```
We are using timestamp from the heartbeat as a reference value even though the actual arrival of heartbeat occurs later when agent is retrying.

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
- removed Heartbeat.timestamp
- updated HeartbeatControllerTest

It closes #1610

### CODE REVIEW

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/config/ConfigProperties.kt [hunk 1]
```diff
@@ -36,7 +36,7 @@ data class ConfigProperties(
     val aptExtraFlags: String = "",
     val adjustResourceOwner: Boolean = true,
     val agentsHeartBeatTimeoutMillis: Long,
-    val heartBeatInspectorInterval: Long,
+    val heartBeatInspectorCron: String,
```
Comment (nulls) [inline save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/config/ConfigProperties.kt:39]: we do not use it from `ConfigProperties`

File: save-orchestrator-common/src/test/kotlin/com/saveourtool/save/orchestrator/controller/heartbeat/HeartbeatControllerTest.kt [hunk 2]
```diff
@@ -382,18 +376,27 @@ class HeartbeatControllerTest {
             whenever(orchestratorAgentService.updateAgentStatus(any()))
                 .thenReturn(emptyResponseAsMono)
         }
-        if (mockAgentStatusesByExecutionId) {
-            whenever(orchestratorAgentService.getAgentStatusesByExecutionId(any()))
+        repeat(mockAgentStatusesByExecutionId) {
```
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/test/kotlin/com/saveourtool/save/orchestrator/controller/heartbeat/HeartbeatControllerTest.kt]: done

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt [hunk 3]
```diff
@@ -183,18 +183,21 @@ class AgentService(
...
-        .getAgentStatusesByExecutionId(executionId)
-        .map { agentStatuses ->
-            log.debug { "For executionId=$executionId agent statuses are $agentStatuses" }
-            // with new logic, should we check only for CRASHED, STOPPED, TERMINATED?
-            agentStatuses.areFinishedOrStopped()
-                .also { areAllAgentsFinishedOrStopped ->
-                    if (areAllAgentsFinishedOrStopped) {
-                        log.debug { "For execution id=$executionId there are finished or stopped agents" }
+    private fun haveAllAgentsFinalStatusByExecutionId(executionId: Long): Mono<Boolean> {
+        log.debug { "Called for executionId=$executionId" }
```
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt]: done

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/config/ConfigProperties.kt (hunk 1 above)
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/config/ConfigProperties.kt:39]: repeat current behavior

File: save-orchestrator-common/src/test/kotlin/com/saveourtool/save/orchestrator/controller/heartbeat/HeartbeatControllerTest.kt [hunk 4]
```diff
@@ -68,12 +69,12 @@ class HeartbeatControllerTest {
             .mutate()
             .responseTimeout(2.seconds.toJavaDuration())
             .build()
+        agentStatusInMemoryRepository.clear()
     }
 
     @AfterEach
     fun cleanup() {
         verifyNoMoreInteractions(orchestratorAgentService)
-        agentStatusInMemoryRepository.clear()
```
Comment (kgevorkyan) [inline save-orchestrator-common/src/test/kotlin/com/saveourtool/save/orchestrator/controller/heartbeat/HeartbeatControllerTest.kt:75]: don't we need to hold this anyway? what if this `cleanup` will be called manually and expect the cleaning?

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt]: actually, it's already removed/reverted -- I added it for debugging purposes only

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kgevorkyan) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt]: oh, sorry

**Pull request merged**